airflow/providers/amazon/aws/hooks/emr.py

```diff
@@ -118,13 +128,13 @@ def create_job_flow(self, job_flow_overrides: dict[str, Any]) -> dict[str, Any]:
                         stacklevel=2,
                     )
                 config = emr_conn.extra_dejson.copy()
-        config.update(job_flow_overrides)
+        config.update(**job_flow_overrides)
```

Comment:
The IDE threw a type error at this line once I added a type hint for `job_flow_overrides` - Prefixing the `**` removed it.

I believe this shouldn't be a breaking change because the tests still pass, thought I'd explicitly point it out as I'd like a second set of eyes on this change.

Reply:
I thought IDE warn because it decide that `RunJobFlowInputRequestTypeDef` not suitable for `dict.update`.

One thing: `job_flow_overrides` might not be `RunJobFlowInputRequestTypeDef`. It is more related to how EMR hook deal with this parameter - described in recent PR https://github.com/apache/airflow/pull/26687

Reply:
I thought IDE warn because it decide that `RunJobFlowInputRequestTypeDef` not suitable for `dict.update`.

One thing: `job_flow_overrides` might not be `RunJobFlowInputRequestTypeDef`. It is more related to how EMR hook deal with this parameter - described in recent PR https://github.com/apache/airflow/pull/26687

Reply:
prefixing with the `**` operator is going to unpack the dict into individual keyword args to the `update` method, **this is not the same** as updating with a dict object (how the previous code was). If you unpack a dict as individual keyword args to update, it puts the constraint on the keys that they must be valid python variable/argument names (i.e. strings). Here's an example that shows things more clearly:
![Screenshot from 2022-10-13 14-36-33](https://user-images.githubusercontent.com/65743084/195715628-a15fdfb3-7510-4d5b-9cbd-19179313cd85.png)

Reply:
@Taragolis:

>I thought IDE warn because it decide that `RunJobFlowInputRequestTypeDef` not suitable for dict.update

Yep, the warning was similar to that. The exact warning was `Expected type 'Mapping[_KT, _VT]', got 'RunJobFlowInputRequestTypeDef'` instead

&nbsp;

>`job_flow_overrides` might not be `RunJobFlowInputRequestTypeDef`

According to the docs in that PR (and confirmed by the [usage](https://github.com/apache/airflow/blob/fb741fd87254e235f99d7d67e558dafad601f253/airflow/providers/amazon/aws/hooks/emr.py#L121-L123) of job_flow_overrides), the result after overriding will be used in the boto3 emr client `run_job_flow` method. So, any values used in `job_flow_overrides` needs to be a subset of the possible values from `RunJobFlowInputRequestTypeDef`.

Please let me know if I'm missing something.

Reply:
@o-nikolas:
Thank you for the example, that helped me understand the difference.

>it puts the constraint on the keys that they must be valid python variable/argument names

I agree, however, I believe this constraint is already present on that variable. Please see the usage [here](https://github.com/apache/airflow/blob/fb741fd87254e235f99d7d67e558dafad601f253/airflow/providers/amazon/aws/hooks/emr.py#L121-L123) - `job_flow_overrides` is used purely to build the `config` dictionary, and that one is used in `run_job_flow(**config)`, hence all keys in there need to be valid identifiers anyway.

If I am mistaken, another option is to replace this line with `config.update(dict(job_flow_overrides))` which removes the constraint on `job_flow_overrrides`.

Can you please let me know how to proceed?

Reply:
@o-nikolas:
Thank you for the example, that helped me understand the difference.

>it puts the constraint on the keys that they must be valid python variable/argument names

I agree, however, I believe this constraint is already present on that variable. Please see the usage [here](https://github.com/apache/airflow/blob/fb741fd87254e235f99d7d67e558dafad601f253/airflow/providers/amazon/aws/hooks/emr.py#L121-L123) - `job_flow_overrides` is used purely to build the `config` dictionary, and that one is used in `run_job_flow(**config)`, hence all keys in there need to be valid identifiers anyway.

If I am mistaken, another option is to replace this line with `config.update(dict(job_flow_overrides))` which removes the constraint on `job_flow_overrides`.

Can you please let me know how to proceed?

Reply:
> So, any values used in job_flow_overrides needs to be a subset of the possible values from RunJobFlowInputRequestTypeDef.
>Please let me know if I'm missing something.

Yeah it might be a subset but separately it not required as `RunJobFlowInputRequestTypeDef`.
So might be better keep annotation as it is: `dict(str, Any)`

Reply:
@dakshin-k 
I don't know EMR well enough to say for sure, but if you're confident all valid values that can be in the dict now or in the future will be valid arguments to a function in python (i.e. stringifiable keys) then feel free to keep what you have (and my gut says you're probably right about that).

Also I do like @Taragolis suggested type hint as well :+1:

Reply:
In additional it might difficult to keep valid definition (keys for specific boto3 client methods).
Due to my past experience is method support specific parameters not only depend on current AWS API but also depends on `boto3` and `botocore` version.

I have been in the position when I need to upgrade `boto3` version to lowest version which support `Concurrecy` in  `run_job_flow` method. Fortunately it was happen couple years ago and couple years ago `boto3` and `botocore` has a bit strange cross dependencies and almost every time use pinned version of `requests`, right not it not a problem.

And second situation `mypy-boto3-<service-name>` should be the same version as `boto3` otherwise it would show incorrect annotation. I don't think it is possible to achieve this so all annotation it is more _nice feature_ rather than actual strict definition of `boto3` client methods